command/node_drain.go

```diff
@@ -259,11 +264,136 @@ func (c *NodeDrainCommand) Run(args []string) int {
 	}
 
 	// Toggle node draining
-	if _, err := client.Nodes().UpdateDrain(node.ID, spec, !keepIneligible, nil); err != nil {
+	meta, err := client.Nodes().UpdateDrain(node.ID, spec, !keepIneligible, nil)
+	if err != nil {
 		c.Ui.Error(fmt.Sprintf("Error updating drain specification: %s", err))
 		return 1
 	}
 
 	c.Ui.Output(fmt.Sprintf("Node %q drain strategy set", node.ID))
+
+	if enable && !detach {
+		if err := monitorDrain(c.Ui.Output, client.Nodes(), node.ID, meta.LastIndex); err != nil {
+			c.Ui.Error(fmt.Sprintf("Error monitoring drain: %v", err))
+			return 1
+		}
+
+		c.Ui.Output(fmt.Sprintf("Node %q drain complete", nodeID))
+	}
+
 	return 0
 }
+
+// monitorDrain monitors the node being drained and exits when the node has
+// finished draining.
+func monitorDrain(output func(string), nodeClient *api.Nodes, nodeID string, index uint64) error {
+	doneCh := make(chan struct{})
+	defer close(doneCh)
+
+	// Errors from either goroutine are sent here
+	errCh := make(chan error, 1)
+
+	// Monitor node changes and close chan when drain is complete
+	nodeCh := make(chan struct{})
+	go func() {
+		for {
+			q := api.QueryOptions{
+				AllowStale: true,
+				WaitIndex:  index,
+			}
+			node, meta, err := nodeClient.Info(nodeID, &q)
+			if err != nil {
+				select {
+				case errCh <- err:
+				case <-doneCh:
+				}
+				return
+			}
+
+			if node.DrainStrategy == nil {
+				close(nodeCh)
+				return
+			}
+
+			// Drain still ongoing
+			index = meta.LastIndex
+		}
+	}()
+
+	// Monitor alloc changes
+	allocCh := make(chan string, 1)
+	go func() {
+		allocs, meta, err := nodeClient.Allocations(nodeID, nil)
+		if err != nil {
+			select {
+			case errCh <- err:
+			case <-doneCh:
+			}
+			return
+		}
+
+		initial := make(map[string]*api.Allocation, len(allocs))
+		for _, a := range allocs {
+			initial[a.ID] = a
+		}
+
+		for {
+			q := api.QueryOptions{
+				AllowStale: true,
+				WaitIndex:  meta.LastIndex,
+			}
+
+			allocs, meta, err = nodeClient.Allocations(nodeID, &q)
+			if err != nil {
+				select {
+				case errCh <- err:
+				case <-doneCh:
+				}
+				return
+			}
+
+			for _, a := range allocs {
+				// Get previous version of alloc
+				orig, ok := initial[a.ID]
+
+				// Update local alloc state
+				initial[a.ID] = a
+
+				msg := ""
+				switch {
+				case !ok:
+					// Should only be possible if response
+					// from initial Allocations call was
+					// stale. No need to output
+
+				case orig.ClientStatus != a.ClientStatus:
+					// Alloc status has changed; output
+					msg = fmt.Sprintf("status %s -> %s", orig.ClientStatus, a.ClientStatus)
+
```

Comment:
Can you add a case for NextAllocation != "" and then make the message, "replaced by allocation %q".